astropy/coordinates/angles/core.py

```diff
@@ -333,6 +333,10 @@ def to_string(
                 if format not in separators:
                     raise ValueError(f"Unknown format '{format}'")
                 sep = separators[format][unit]
+            elif sep == "hms":
+                unit = u.hourangle
+            elif sep == "dms":
+                unit = u.degree
```

Comment:
This works too, though I note that the `if decimal` branch becomes redundant with the default value in the `dict.get`. To a certain extent I feel like this is obfuscating the actual logic, actually equivalent to:
```python
if unit is None:
    if sep == "dms":
        unit = u.degree
    elif sep == "hms":
        unit = u.hourangle
    else:
        unit = self.unit
else:
    unit = self._convert_unit_to_angle_unit(u.Unit(unit))
```
I'll push this dumb-simple version !

Reply:
This would effectively treat the first two cases as `decimal=None`, which is the only valid combination with those values of `sep`, right? But illegal settings of `decimal` are not explicitly checked here (probably still would raise below in `else: if sep != "fromunit": ...` – perhaps this should still be done explicitly?

Reply:
Truthy values for `decimal`, combined with `unit == u.degree` or `unit == u.hourangle`, are already explicitly disallowed.

https://github.com/astropy/astropy/blob/1f45d3817b90ef60e92e7d49252a07cd7c6a24c4/astropy/coordinates/angles/core.py#L343-L347

I started this comment thinking that adding more `raise` statement would be redundant but then I needed a couple minutes of going through the code to make sure that it was true... which I now think is reason enough to add this "redundant" check. At least it'll make the logic easier to read through, and I can also make a slightly confusing error message !

Reply:
That was the raise "below" I referred to, but was not sure if that would catch all cases. I think it does, but looks still better to me to catch it early on as with your last commit.